Description: Do not free packets in ndpiReader because they may used at a later
 time. This creates a memory leak, but at least permits consistent test results.
Author: Ludovico Cavedon <cavedon@debian.org>
Bug: https://github.com/ntop/nDPI/issues/657

Index: ndpi/example/ndpiReader.c
===================================================================
--- ndpi.orig/example/ndpiReader.c
+++ ndpi/example/ndpiReader.c
@@ -2489,7 +2489,9 @@ static void ndpi_process_packet(u_char *
   if(memcmp(packet, packet_checked, header->caplen) != 0)
     printf("INTERNAL ERROR: ingress packet was modified by nDPI: this should not happen [thread_id=%u, packetId=%lu, caplen=%u]\n",
 	   thread_id, (unsigned long)ndpi_thread_info[thread_id].workflow->stats.raw_packet_count, header->caplen);
-  free(packet_checked);
+  /* Pointers ot packets are stored inside flow data structures, and may be
+   * referenced when printing stats. We cannot deallocate the packet here. */
+  /* free(packet_checked); */
 
   if((pcap_end.tv_sec-pcap_start.tv_sec) > pcap_analysis_duration) {
     int i;
@@ -3298,6 +3300,11 @@ int orginal_main(int argc, char **argv)
 	     "* In this demo we have implemented only some basic features\n"
 	     "* just to show you what you can do with the library. Feel \n"
 	     "* free to extend it and send us the patches for inclusion\n"
+	     "------------------------------------------------------------\n"
+             "* WARNING: This version has been patch to fix an invalid\n"
+             "* memory reference. A side effect of the fix is that all\n"
+	     "* packets are kept in memory. Therefore, it cannot be used\n"
+	     "* for large amounts of traffic.\n"
 	     "------------------------------------------------------------\n\n");
 
       printf("Using nDPI (%s) [%d thread(s)]\n", ndpi_revision(), num_threads);
